File: 322/coin-change.java
class Solution {
    //啊，很亂，重寫啦！
    public int coinChange(int[] coins, int amount) {
        int [][] table = new int[coins.length+1][amount+1];
        for(int i=0; i<=coins.length; i++){
            for(int j=0; j<=amount; j++){
                table[i][j] = -1;
            }
        }
        for(int k=0; k<=coins.length; k++){
            table[k][0]=0;
        }
        Arrays.sort(coins);
        for(int i=1; i<=coins.length; i++){
            for(int k=1; k<=amount; k++){
                int prev = k - coins[i-1];
                if(prev<0){
                    table[i][k] = table[i-1][k];
                    continue;
                }
                int a = table[i-1][k];
                int b = table[i][prev];
                if(a==-1 && b==-1) continue;
                else if(a==-1) table[i][k] = b+1;
                else if(b==-1) table[i][k] = a;
                else table[i][k] = min(a, b+1);
            }
        }
        return table[coins.length][amount];
    }
    int min(int a, int b) { //which is smaller
        if(a<b) return a;
        else return b;
    }
}

File: 322/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        table = [-1]*(amount+1) # table[i] 達到i元, 答案是 table[amount]
        table[0] = 0 # 0個銅板，便能達到0元
        coins = sorted(coins) # 希望從小到大的順序來做
        for k in range(len(coins)):
            c = coins[k]
            for i in range(amount+1):
                if i>=c and table[i-c]!=-1: # 能更新，且前面有值，
                    if table[i]==-1 or table[i-c]+1<table[i]: 
                        # 還沒有任何值 or 有更好的值
                        table[i] = table[i-c]+1
        print(table)
        return table[amount]
# case 99/189: [186,419,83,408] 6249
